### ISSUE TITLE
Throws during pure calls to library functions

### ISSUE THREAD

**Issue by pelsasser** [sc-forks/solidity-coverage#234]
Hello sc team,

Recently we have attempted to upgrade to newer version of solidity as well as solidity-coverage and found some issues that took a very long time to track down.

I have created a very minimal example to show the issue which can be found [here](https://github.com/MARKETProtocol/solidity-coverage-bug).

Essentially, a `.call()` to a `pure` library function is throwing `Error: VM Exception while processing transaction: revert` in the coverage environment.

If you take a look at the repo, the instructions to reproduce are there.

According to the docs of solidity 0.4.23 this should be possible and has worked in the past 

> Library functions can only be called directly (i.e. without the use of DELEGATECALL) if they do not modify the state (i.e. if they are view or pure functions), because libraries are assumed to be stateless

please let me know what other information I can help provide to get this resolved.  I think the repo should help bring out the issue pretty clearly.

Thanks for all you guys do!

**Comment by cgewecke** [comment]
Hi @pelsasser. Thanks so much for opening. Quite sorry because this issue is known and should have been documented. It was discovered by Zeppelin about a month ago (more detail in [this thread](https://github.com/OpenZeppelin/openzeppelin-solidity/pull/876#issuecomment-379621933))

The underlying issue is that `solidity-coverage` rewrites the contract ABIs of pure/view functions in order to sneak event-based instrumentation into the contracts during compilation. It then swaps the original ABIs back into the contract artifacts to get `truffle` to call the functions correctly. (e.g. as calls, not transactions). 

Since `0.4.20`, in response to Parity's wallet being deleted, solc has added protections  (at the bytecode level) that prevent non-view library methods from being called directly. TLDR, the pure library functions are being compiled as non-view functions and this triggers the error.

A work-around to call these library methods through a mock contract that uses the library. 

+ an example at Zeppelin [here](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/mocks/MerkleProofWrapper.sol)
+ mocks must be located in the `contracts` directory (not `test`). You can put them in a `mocks` folder.
+ mock contracts themselves can be excluded from coverage using the `skipFiles` option in `.solcover.js`, e.g 
  ```javascript
  skipFiles: [
        'lifecycle/Migrations.sol', // <-- single file
        'mocks' // <-- whole folder
    ]
  ```
+ if you are publishing your contracts to `npm` you'll probably want to `.npmignore` your mocks folder

**Comment by pelsasser** [comment]
@cgewecke - thanks for the quick reply.

I had previously seen that thread, but didn't quite put it all together that is was directly tied to the issue I was experiencing.  Good to know there is a workaround!

I will get the mocks coded up!

**Comment by djudjuu** [comment]
Thanks for this helpful instruction. I'd like to add that I add to add the ```port: 6545``` line to my ```.solcover.js``` to make it work.

**Comment by Agusx1211** [comment]
With the new _emitFreeLogs_ flag on [ethereumjs-vm](https://github.com/ethereumjs/ethereumjs-vm), it should be possible to instantiate testrpc setup to permit events on staticcalls, it may be required to add the parameter to test-rpc or ganache.

**Comment by cgewecke** [comment]
Fixed with `0.7.0`
